emc/model.py

```diff
@@ -0,0 +1,216 @@
+                ExponentialIsotropicModel,
+            )
+
+            param = {
+                "isotropic": ExponentialIsotropicModel,
+            }
+
+        elif model.lower().startswith("tensor"):
+            Model = TensorModel
+
+        elif model.lower().startswith("dki"):
+            Model = DKIModel
+
+        else:
+            raise NotImplementedError(f"Unsupported model <{model}>.")
+
+        param.update(kwargs)
+        return Model(gtab, **param)
+
+
+class TrivialB0Model:
+    """
+    A trivial model that returns a *b=0* map always.
+
+    Implements the interface of :obj:`dipy.reconst.base.ReconstModel`.
+    Instead of inheriting from the abstract base, this implementation
+    follows type adaptation principles, as it is easier to maintain
+    and to read (see https://www.youtube.com/watch?v=3MNVP9-hglc).
+
+    """
+
+    __slots__ = ("_S0",)
+
+    def __init__(self, gtab, S0=None, **kwargs):
+        """Implement object initialization."""
+        if S0 is None:
+            raise ValueError("S0 must be provided")
+
+        self._S0 = S0
+
+    def fit(self, *args, **kwargs):
+        """Do nothing."""
+
+    def predict(self, gradient, **kwargs):
+        """Return the *b=0* map."""
+        return self._S0
+
+
+class TensorModel:
+    """A wrapper of :obj:`dipy.reconst.dti.TensorModel."""
+
+    __slots__ = ("_model", "_S0", "_mask")
+
+    def __init__(self, gtab, S0=None, mask=None, **kwargs):
+        """Instantiate the wrapped tensor model."""
+        from dipy.reconst.dti import TensorModel
+
+        self._S0 = None
+        if S0 is not None:
+            self._S0 = np.clip(
+                S0.astype("float32") / S0.max(),
+                a_min=1e-5,
+                a_max=1.0,
+            )
+        self._mask = mask
+        if mask is None and S0 is not None:
+            self._mask = self._S0 > np.percentile(self._S0, 35)
+
+        if self._mask is not None:
+            self._S0 = self._S0[self._mask.astype(bool)]
+
+        kwargs = {
+            k: v
+            for k, v in kwargs.items()
+            if k
+            in (
+                "min_signal",
+                "return_S0_hat",
+                "fit_method",
+                "weighting",
+                "sigma",
+                "jac",
+            )
+        }
+        self._model = TensorModel(gtab, **kwargs)
+
+    def fit(self, data, **kwargs):
+        """Call model's fit."""
+        if self._mask is not None:
+            data = data[self._mask, ...]
+        self._model = self._model.fit(data)
+
+    def predict(self, gradient, step=None, **kwargs):
+        """Propagate model parameters and call predict."""
+        predicted = np.squeeze(
+            self._model.predict(
+                _rasb2dipy(gradient),
+                S0=self._S0,
+                step=step,
+            )
+        )
+        if predicted.ndim == 3:
+            return predicted
+
+        retval = np.zeros_like(self._mask, dtype="float32")
```

Comment:
It seems here you are making the assumption that `self._mask is not None` as you flatten in this case the data given to the fit, is that correct?

Reply:
Yes, I believe it was in my latest commit, I generate a mask by thresholding the data if no mask is given. Therefore, it should not be `None` at this point.